Dev/actions updates

alexjhawk opened the pull request.

alexjhawk commented:
@it-hms @TomKimsey,

I removed the commit(s) tested against the actions update. They were:
1. https://github.com/hms-networks/sc-java-maven-starter-project/commit/a75addcd2f2b9cdba9f5ec24952563b11d4142f9
2. https://github.com/hms-networks/sc-java-maven-starter-project/commit/9f155fa5c86c962c6502e2436b41e0fe0b27e628

Notably, this change to the action does duplicate some of the functionality already given by the general Google Java format tool/action. It expands further by also checking additional Google Java Format conventions, such as variable naming conventions. It also provides a more friendly comment on the exact formatting concern raised/detected (see above).

As discussed, though, this is only one step in moving toward improving the system. There is plenty of room for investigation into and implementation of a system (such as a linter), which would universally prevent formatting issues from being committed to begin with.

TomKimsey commented:
@alexjhawk is there anyway for the changes requested by github-actions[bot] to be dismissed on push? Its doesnt really look clean when there are unresolved changes requested even though those changes are stale.

alexjhawk commented:
> @alexjhawk is there anyway for the changes requested by github-actions[bot] to be dismissed on push? Its doesnt really look clean when there are unresolved changes requested even though those changes are stale.

I'll look into this -- I believe there are commands available at a minimum, but there likely is a way to configure it automatically if desired.

alexjhawk commented:
> > @alexjhawk is there anyway for the changes requested by github-actions[bot] to be dismissed on push? Its doesnt really look clean when there are unresolved changes requested even though those changes are stale.
> 
> I'll look into this -- I believe there are commands available at a minimum, but there likely is a way to configure it automatically if desired.

It doesn't seem to be natively possible, but I can extend the action so that it automatically resolves any 'conversations' that are 'outdated' from a specific user (the bot), before performing the actual check. This would clean up the review comments/conversations, and reviewdog would review again/comment as necessary. What are your thoughts on that approach?

TomKimsey commented:
> > > @alexjhawk is there anyway for the changes requested by github-actions[bot] to be dismissed on push? Its doesnt really look clean when there are unresolved changes requested even though those changes are stale.
> > 
> > 
> > I'll look into this -- I believe there are commands available at a minimum, but there likely is a way to configure it automatically if desired.
> 
> It doesn't seem to be natively possible, but I can extend the action so that it automatically resolves any 'conversations' that are 'outdated' from a specific user (the bot), before performing the actual check. This would clean up the review comments/conversations, and reviewdog would review again/comment as necessary. What are your thoughts on that approach?

Yeah, i think i would prefer if those were just deleted as if the problem still exists they will be created again. Resolving is fine though, benefit of that is it maintains a history of the issue. Not that keeping a history of those errors is important.

alexjhawk commented:
The above discussion also brings up the topic of access tokens. It looks like the auto-resolve piece would require a token that has the ability to access the PR and resolve a conversation. This is probably a non-issue with a public repo, but private repositories could be a concern. That may just be a general concern overall as well considering whether the bot has necessary permissions on a private repo.

alexjhawk commented:
> > > > @alexjhawk is there anyway for the changes requested by github-actions[bot] to be dismissed on push? Its doesnt really look clean when there are unresolved changes requested even though those changes are stale.
> > > 
> > > 
> > > I'll look into this -- I believe there are commands available at a minimum, but there likely is a way to configure it automatically if desired.
> > 
> > 
> > It doesn't seem to be natively possible, but I can extend the action so that it automatically resolves any 'conversations' that are 'outdated' from a specific user (the bot), before performing the actual check. This would clean up the review comments/conversations, and reviewdog would review again/comment as necessary. What are your thoughts on that approach?
> 
> Yeah, i think i would prefer if those were just deleted as if the problem still exists they will be created again. Resolving is fine though, benefit of that is it maintains a history of the issue. Not that keeping a history of those errors is important.

Long term, the goal would be to integrate this on the Maven front as well, so that a user would be forced to use the proper formatting to begin with. This action would almost immediately become redundant (but not useless or worth removing) following that change.

alexjhawk commented:
> > > > @alexjhawk is there anyway for the changes requested by github-actions[bot] to be dismissed on push? Its doesnt really look clean when there are unresolved changes requested even though those changes are stale.
> > > 
> > > 
> > > I'll look into this -- I believe there are commands available at a minimum, but there likely is a way to configure it automatically if desired.
> > 
> > 
> > It doesn't seem to be natively possible, but I can extend the action so that it automatically resolves any 'conversations' that are 'outdated' from a specific user (the bot), before performing the actual check. This would clean up the review comments/conversations, and reviewdog would review again/comment as necessary. What are your thoughts on that approach?
> 
> Yeah, i think i would prefer if those were just deleted as if the problem still exists they will be created again. Resolving is fine though, benefit of that is it maintains a history of the issue. Not that keeping a history of those errors is important.

Deletion is also an option: https://github.com/Ardiannn08/resolve-outdated-comment

Same issue regarding access tokens though, which may be a problem with private repositories

TomKimsey commented:
Ill be out for the next two days. Ill approve for now given the concerns about access tokens.

alexjhawk commented:
> Ill be out for the next two days. Ill approve for now given the concerns about access tokens.

I'll likely just merge this change to the starter project soon/later this week, and work on addressing the concerns with access tokens/conversation resolution separately if needed.

alexjhawk commented:
Pull request merge is blocked by outstanding requested changes. I will need to work on this action to ensure it does not hinder the pull request process (in addition to access token investigation).

alexjhawk commented:
Significant progress has been made, outlined below:

- [ ] Fix the major 'bug' that could cause PR to be unable to merge due to specific repository protection settings.
  - Added PR approval to action if reviewdog succeeds
- [ ] Lessen dependence on third party actions
  - This turned in to many notable improvements, but ultimately, we now directly use reviewdog, and invoke Checkstyle through Maven.
- [ ] Embed matching Checkstyle functionality into Maven
  - I was planning on doing this later, but to address both of the above checkboxes/TODOs, the official Apache Maven plugin for checkstyle was a natural fit. 
  - The GitHub action now invokes Checkstyle through our Maven `pom.xml` and passes those results to reviewdog.
  - This ensures consistency in reporting and eliminates the need for a standalone Checkstyle setup in the GitHub action.
- [ ] Address possible concerns with private repository access
  - [ ] Checkstyle access to private repo code
    - By using Checkstyle through Maven, this is no longer a concern as no external/3rd party action is required
  - [ ] Reviewdog access to private repo code
    - Some concern alleviated by 'installing' and directly using reviewdog, testing pending
    - Comment(s) added in action YML about locations to change token(s)
    - Comment(s) added in action YML about possible Reviewdog GitHub app support which would solve access concern
  - [ ] PR Approve access to private repo
    - Currently using an approved GitHub action for PR approval, which uses GitHub OctoKit REST API internally
    - Not sure if the access token given to the GitHub Action by default has access to do this, testing pending
- [ ] Address outdated comments from reviewdog. These should be deleted or resolved once the concern has been addressed/fixed to avoid PR noise

alexjhawk commented:
Regarding Maven/IDE integration, here is an example of that in action:

![image](https://github.com/hms-networks/sc-java-maven-starter-project/assets/51409174/d5dbbc57-e09e-45bf-b1fc-c5b2ddfca77b)

TomKimsey commented:
Yeah, i like this new direction. 

What does installing reviewdog entail?

alexjhawk commented:
> Yeah, i like this new direction.
> 
> What does installing reviewdog entail?

I believe you just need to `install` by giving the app permission to access org repositories.
